Docker Image documentation; updated and transferred from separate repository Readme (#74)
payara/Payara-Community-Documentation: docs/modules/ROOT/pages/documentation/ecosystem/docker-micro-usage.adoc

== Usage

=== Quick start

To start the docker container and run Payara Micro:

....
docker run -p 8080:8080 payara/micro
....

It runs Payara Micro without any applications, therefore accessing the HTTP server bound to port 8080 will just return HTTP code 404 - Not Found.

You need to add your applications to the container and deploy them.

It is always recommended to specify the tag name of the Docker Image that you like to run. Otherwise, you are running the latest version and might encounter some unwanted effects due to changes between versions.

....
docker run -p 8080:8080 payara/micro:{currentVersion}
....

=== JDK Version

_Since Payara 5.2020.3_

The Docker Images are available in 2 different flavours. The one is running on JDK 8 and the other has JDK 11 on board.

> The command used in the Quick start above, will start the Zulu JDK 8 version. If you need the Zulu JDK 11 version, you need to add the `-jdk11` suffix to the image name.

Author: It is not a good practice to use the LATEST (or jdk11 tag here) as you do not have control over the moment when a newer version of Payara is used). That is the reason i isn't mentioned.

The following example Dockerfile will build an image that deploys `myapplication.war` when Payara Micro starts:

....
FROM payara/micro:{currentVersion}

COPY myapplication.war $DEPLOY_DIR
....

==== *Run from a maven repository*

If your application is already in a maven repository, you can run it with Payara Micro in the docker very easily. Payara Micro knows how to download an artifact from a maven repository and run it directly. You just need to provide an alternative entry point

The following command runs Payara Micro in the docker image and runs an application stored in a maven repository. The application group is `fish.payara.examples`, artifact name is `my-application`, and version is `1.0-SNAPSHOT`. The maven repository is available on host `172.17.0.10`:

....
docker docker run -p 8080:8080 payara/micro:{currentVersion} \
 --deployFromGAV "fish.payara.examples:my-application:1.0-SNAPSHOT" \
 --additionalRepository https://172.17.0.10/content/repositories/snapshots
....

==== *Define context root*

_Since Payara Micro 5.191_

> The context root of the deployed application(s) is determined based on the file name of the artifact which is deployed. Form Payara Micro version 5.191 on, you can explicitly define the value of the context root using the `--contextroot` option, including deploying the application on the root itself ( `/`, specify *ROOT* as option value).

Reviewer: "From version" not needed, it's already specified above.

```suggestion
The context root of the deployed application(s) is determined based on the file name of the artifact which is deployed. You can explicitly define the value of the context root using the `--contextroot` option, including deploying the application on the root itself ( `/`, specify *ROOT* as option value).
```

Some examples of how it can be used :

....
docker run -p 8080:8080 \
 -v ~/payara-micro/applications:/opt/payara/deployments payara/micro:{currentVersion}  \
 --deploymentDir /opt/payara/deployments  \
 --contextroot myRoot
....

Deploys the (first) application within the mounted directory _~/payara-micro/applications_ under the root context _/myRoot_.

....
docker run -p 8080:8080 \
 -v ~/payara-micro/applications:/opt/payara/deployments \
 payara/micro:{currentVersion} \
 --deploy /opt/payara/deployments/myapplication.war \
 --contextroot ROOT
....

Deploys the _myapplication.war_ to the root.

....
FROM payara/micro:{currentVersion}

COPY myapplication.war $DEPLOY_DIR

CMD ["--deploymentDir", "/opt/payara/deployments", "--contextroot", "my"]
....

Creates a new Docker Image which results in the deployment of the _myapplication.war_ under the root context _/my_.

== Speeding up Boot by disabling the Data Grid

> By default the image has the Data Grid enabled and instances of Payara Micro will form a Data Grid if they can see each other over multicast. Typically this will occur on a single host.

Reviewer: The last sentence doesn't make any sense to me.

```suggestion
By default the image has the Data Grid enabled and instances of Payara Micro will form a Data Grid if they can see each other over multicast. 
```

> To speed up boot time the Data Grid can be disabled by modifying the CMD to add `--nocluster`

Reviewer: ```suggestion
If you don't intend to form a cluster and want to speed up boot time, you can disable Data Grid by modifying the CMD to add `--nocluster`
```

== Configuration

=== Environment Variables

The following environment variables are available to configure Payara Micro. 

> * `MEM_MAX_RAM_PERCENTAGE` - Value for the JVM parameter `-XX:MaxRAMPercentage` that indicate the percentage of memory assigned to the container that can be used by the Java Process. By default this is _70_.

Reviewer: ```suggestion
* `MEM_MAX_RAM_PERCENTAGE` - Value for the JVM parameter `-XX:MaxRAMPercentage` that indicates the percentage of memory assigned to the container that can be used by the Java Process. By default this is _70_.
```
